src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Formatting/Passes/New/CSharpFormattingPass.cs

```diff
@@ -0,0 +1,146 @@
+
+                // First up, we take the indentation from the formatted file, and add on the Html indentation level from the line info, and
+                // replace whatever was in the original file with it.
+                var htmlIndentString = context.GetIndentationLevelString(lineInfo.HtmlIndentLevel);
+                var indentationString = formattedCSharpText.ToString(new TextSpan(formattedLine.Start, formattedIndentation))
+                    + htmlIndentString
+                    + lineInfo.AdditionalIndentation;
+                formattingChanges.Add(new TextChange(new TextSpan(originalLine.Start, originalLineOffset), indentationString));
+
+                // Now we handle the formatting, which is changes to the right of the first non-whitespace character.
+                if (lineInfo.ProcessFormatting)
+                {
+                    // The offset and length properties of the line info are relative to the indented content in their respective documents.
+                    // In other words, relative to the first non-whitespace character on the line.
+                    var originalStart = originalLine.Start + originalLineOffset + lineInfo.OriginOffset;
+                    var length = lineInfo.FormattedLength == 0
+                        ? originalLine.End - originalStart
+                        : lineInfo.FormattedLength;
+                    var formattedStart = formattedLine.Start + formattedIndentation + lineInfo.FormattedOffset;
+                    formattingChanges.Add(new TextChange(new TextSpan(originalStart, length), formattedCSharpText.ToString(TextSpan.FromBounds(formattedStart, formattedLine.End - lineInfo.FormattedOffsetFromEndOfLine))));
+
+                    if (lineInfo.CheckForNewLines)
+                    {
+                        Debug.Assert(lineInfo.FormattedLength == 0, "Can't have a FormattedLength if we're looking for new lines. The logic is incompatible.");
+                        Debug.Assert(lineInfo.FormattedOffsetFromEndOfLine == 0, "Can't have a FormattedOffsetFromEndOfLine if we're looking for new lines. The logic is incompatible.");
+
+                        // We assume Roslyn won't change anything but whitespace, so we can just apply the changes directly, but
+                        // it could very well be adding whitespace in the form of newlines, for example taking "if (true) {" and
+                        // making it run over two lines, or even "string Prop { get" and making it span three lines.
+                        // Since we assume Roslyn won't change anything non-whitespace, we just keep inserting the formatted lines
+                        // of C# until we match the original line contents.
+                        while (!FormattingUtilities.ContentEqualIgnoringWhitespace(changedText, originalStart, originalLine.End, formattedCSharpText, formattedStart, formattedLine.End))
```

Comment:
I have a hunch this could be improved but my brain doesn't have a way yet. This is a note for future me to take another look if I can